Add `xftCharExists` binding and add support for multiple fonts

iliayar opened the pull request:
Mostly move implementation from xmobar's [MinXft](https://github.com/jaor/xmobar/blob/master/src/Xmobar/X11/MinXft.hsc). Now  we are able to remove MinXft from https://github.com/xmonad/xmonad-contrib/pull/646.

Functions with prime is an alternative for working with multiple fonts

iliayar committed:
Add support for multiple fonts

Mostly move implementation from xmobar's MinXft, but rework
calculating chunks extents.

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +245,69 @@ xftTextExtents d f string =
+                  XftColor ->
+                  [XftFont] ->
+                  Integer ->
+                  Integer ->
```
Why use `Integer` here instead of `Int`?

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +245,69 @@ xftTextExtents d f string =
+    -- FIXME: When string is empty xftTextExtents' crashes(and
+    -- maybe some other functions) because of empty list of chunks
+    -- they get. Empty font lists are excluded by the above hand, so
+    -- we can take any(beacuse string is empty anyway).
```
I would imagine we could just catch that exception in `getOffsets` and return an empty list there instead.  `xftTextExtents'` would then need to be modified to return an `IO (Maybe XGlyphInfo)`, but that might be for the best anyways.  Or, even better yet, modify `xftTextExtents` to return a `Maybe` instead of throwing an exception.

EDIT: Indeed, it seems like there is [at least one open issue](https://github.com/xmonad/xmonad/issues/194) regarding `xftTextExtends` that sounds very bad.  Would be nice to drive-by close this :)

iliayar replied:
I think the issue was only in `xftStringExtents'`, which is now returning the list of chunks. So now it's ok for `getChunks` to return an empty list as it is used only by `xftStringExtents`

slotThe replied:
I think there is still definitely an issue in `xftTextExtents` (c.f. [here](https://github.com/xmonad/xmonad/issues/194#issuecomment-509674902)), but you're right that this is perhaps not (nor should it be) the objective of this PR to fix this (though we should definitely do that at some point)

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +245,69 @@ xftTextExtents d f string =
+    runs [] = []
+    runs (x:xs) = let (h, t) = span (==x) xs in (x, length h + 1) : runs t
+    split [] _ = []
+    split ((x, c):xs) s = let (h, t) = splitAt c s in (x, h) : split xs t
```
I find this logic a bit confusing here.  I'd write something like

```haskell
let splits = map (\ll -> (fst (head ll), map snd ll)) -- or even map (fst . head &&& map snd)
           . groupBy ((==) `on` fst)
           $ zip glyphs s
```

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +245,69 @@ xftTextExtents d f string =
+    -- we can take any(beacuse string is empty anyway).
+    getFonts _ (ft:_) s@[] = return [(ft, s)]
+    getFonts _ [ft] s = return [(ft, s)]
+    getFonts d fonts@(ft:_) s = do
```
These passthroughs of `d` aren't necessary; you can just use `disp`

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +245,69 @@ xftTextExtents d f string =
+xftTextExtents' :: Display -> [XftFont] -> String -> IO XGlyphInfo
+xftTextExtents' d fs string = do
+    chunks <- getChunks d fs string
+    let (_, _, gi, _, _) = last chunks
```
Wouldn't it be better to return a list here and let the caller decide what to do with it?  Just arbitrarily picking the font of the last chunk seems unintuitive to me

slotThe reviewed (commented):
Thanks!

I'm not a huge fan of the names for the functions though; I would prefer something more descriptive.  E.g., instead of `xftfont_ascent'` something like `xftfonts_max_ascent`

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -190,14 +193,18 @@ foreign import ccall "XftFontClose"
 
-xftfont_descent' :: [XftFont] -> IO Int
-xftfont_descent' = (fmap maximum) . (mapM xftfont_descent)
+xftfont_max_ascent :: [XftFont] -> IO Int
+xftfont_max_ascent = (fmap $ foldIntNotEmpty max 0) . (mapM xftfont_ascent)
```
Instead of `0` being a vague "something went wrong" value, I would push the responsibility to check this onto the caller, using `NonEmpty XftFont -> IO Int` here (alternatively we could return a `Maybe` or something)

slotThe commented:
> ascent, descent, height now return Nothing on empty list
>   Due to ghc 7.10.3 has no `NonEmpty`

Mh, interesting.  I would almost say then to just drop support for these kinds of ancient compiler versions; we shouldn't have to compromise the API (too much) for that

geekosaur commented:
Are there any other consumers of X11-xft who would be affected by that? `packdeps.haskellers.com/reverse` doesn't seem to think there are many dependents, so maybe not.

slotThe commented:
> Are there any other consumers of X11-xft who would be affected by that? `packdeps.haskellers.com/reverse` doesn't seem to think there are many dependents, so maybe not.

Yeah the only active dependency I could find was... us :)  I suppose we might want to bump to 0.4 or something but other than that I don't see why we wouldn't do this (it's not like the X11 situation where we just had to rename a flag to be compatible with 7.10)

iliayar commented:
So, are you going to drop 7.10.3? Then I can revert to `NonEmpty` variant.

slotThe committed:
Don't test against GHC 7.10.3

We would like to leverage Data.List.NonEmpty for some of this code;
e.g., to definitely (modulo exceptions) return text ascents when given
some non-empty list of fonts.  However, the module only made it into
base in version 4.9, which was released after 7.10.x's time[1].

Dropping compatibility with such an old compiler version—especially
since we're the only active dependency[2]—seems acceptable.

[1]: https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history
[2]: https://packdeps.haskellers.com/reverse/X11-xft

slotThe commented:
> So, are you going to drop 7.10.3? Then I can revert to `NonEmpty` variant.

I just pushed the relevant commit to your branch, should be fine now

slotThe commented on Graphics/X11/Xft.hsc:
```diff
@@ -219,6 +249,58 @@ xftTextExtents d f string =
+
+-- | Like 'xftTextExtents' but for multiple fonts.  Return the extents
+-- for all fonts that are needed to draw the given string.
+xftTextExtentsChunks :: Display -> [XftFont] -> String -> IO [XGlyphInfo]
+xftTextExtentsChunks d fs string = map (\(_, _, gi, _, _) -> gi) <$> getChunks d fs string
```
I think this may be a bit confusing since we could have the same glyphinfo in multiple places in that list (think "string *unicode_symbol* string") but no indication of which parts of the text it belongs to.  I suppose we could either return something like `[(String, XGlyphInfo)]` or we could nub the list (probably easiest by deriving `Eq` on `XftFont` and then calling `nubBy` before `map`).

iliayar replied:
I just mention that `getOffsets` in `getChunks` actually computes wrong `XGlyphInfo` for each independent substring. It accumulates the width and height for whatever reason. Or maybe I misunderstood the semantic of `XGlyphInfo` fields. So I think I could work on that, and then just return `[(String, XGlyphInfo)]` in `xftTextExtentsChunks`

slotThe replied:
Well, font fallback for xmobar certainly works, which makes this very weird indeed.  I'm not so sure about what `XGlyphInfo` is really supposed to represent either—documentation seems rather sparse.

@jaor since this code originally came from xmobar (and according to the module docs from you!) can you shed any light on this?  I hope it's not too indecent to ask about code from like 7 years ago :)

iliayar replied:
In xmobar it makes sense, while `xftTextEtentsChunks` returns the last, which contains info about all chunks in one such struct

slotThe replied:
Ah indeed, I forgot about `xftDrawStringFallback` not using this information at all.  I guess what we should really be doing then is to return something like `IO (XGlyphInfo, [(XftFont, String, Int, Int)])`, right?

iliayar replied:
Actually, we can make two separate ones, `xftTextExtentsChunks` to return `[(String, XftFont, XGlyphInfo)]` (we don't need xo, yo in this tuple because we'll store it in `XGlyphInfo`) and something like `xftTextAccumulativeExtents` to return just `XGlyphInfo` like xmobar does for the main use case.

Then I need to reimplement `getOffsets`. Also, I think we can get rid of xo, yo in tuple at all, because we store it in `XGlyphInfo` anyway. So I need to play with it a bit.

slotThe replied:
At this point we may well just drop the idea of `xftTextExtentsChunks` and just expose `xftTextAccumulativeExtents`.  I'm not sure the fomer has any uses, I just immediately thought of it when I saw that call to `last`

slotThe replied:
@iliayar LGTM, feel free to squash this down and then we can merge

slotThe reviewed (approved):
This looks good to me now!  I just pushed a minor nit fixup.  Other than that there is but a single comment I still have, after that we can squash and merge

slotThe commented:
Thanks!

slotThe committed:
Update CHANGES.md

liskin reviewed (commented):
Can we bump the `base` dependency then? We forgot to do it in X11 and then someone complained that it doesn't build, because cabal didn't know it shouldn't use the version that no longer works with GHC <8.

(this is a comment for the https://github.com/xmonad/X11-xft/pull/2/commits/81c54eb20e08b8a7e9cb2ab1db220af041b43e2b commit, GitHub apparently doesn't let me comment on individual commits post-merge.)

slotThe commented:
Good point!  Fixed by 6cc5fd928cc932728f3395242b6c8115a481f675
